packages/file-search/src/node/file-search-service-impl.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { FileSearchService } from '../common/file-search-service';
+import * as fs from "fs";
+import * as path from "path";
+import URI from '@theia/core/lib/common/uri';
+import { FileUri } from '@theia/core/lib/node';
+import * as fuzzy from 'fuzzy';
+import { injectable, inject } from 'inversify';
+import { ILogger } from '@theia/core';
+import { findContainingGitIgnore, NO_IGNORE, GitIgnore, findGitIgnore } from './git-ignore';
+
+type Filtered = (simpleName: string, fullPath: string) => boolean;
+
+
+@injectable()
+export class FileSearchServiceImpl implements FileSearchService {
+
+    constructor( @inject(ILogger) private logger: ILogger) { }
+
+    async find(uri: string, searchPattern: string, options?: FileSearchService.Options): Promise<string[]> {
+        const basePath = FileUri.fsPath(new URI(uri));
+        const opts = {
+            fuzzyMatch: true,
+            limit: Number.MAX_SAFE_INTEGER,
+            useGitignore: true,
+            defaultIgnorePatterns: [
+                '^.git$'
+            ],
+            ...options
+        };
+        const result: string[] = [];
+        const limitReached = new Error("limit reached");
+        const globalFiltered = this.getFiltered(basePath, opts);
+        const gitignore = opts.useGitignore ? findContainingGitIgnore(basePath) : NO_IGNORE;
```

Comment:
about findContainingGitIgnore:

  - It's not taking into account not only nested repos but repos on the same level for example
  workspace/.git
  workspace/repo/.git
  workspace/repo2/.git

  We're searching file: repo/something/foo.txt but we're reading only workspace/.gitignore,  not repo/.gitignore

  The git extension could tell you those. hence it's value.

  - I think the current way it is done would work if the basePath was the result filePath and not the workspace, but this would make it check for a lot of directories...

Reply:
It's not true. It respects nested git repos and gitignore.

Reply:
How so ?? It starts at workspace then goes up

Reply:
From your tests it seems to work.. I'll debug a bit to understand this properly...